src/Service/ProducerEventProcessor.cpp

```diff
@@ -0,0 +1,405 @@
+
+  CaptureEventBuffer* capture_event_buffer_;
+
+  std::atomic<uint64_t> callstack_id_counter_;
+  std::atomic<uint64_t> string_id_counter_;
+  std::atomic<uint64_t> tracepoint_id_counter_;
+
+  InternedCache<std::vector<uint64_t>> callstack_cache_;
+  InternedCache<std::string> string_cache_;
+  InternedCache<std::pair<std::string, std::string>> tracepoint_cache_;
+
+  // These are mapping InternStrings and Callstacks from producer ids
+  // to client ids:
+  // <producer_id, producer_string_id> -> client_string_id and
+  // <producer_id, producer_callstack_id> -> client_callstack_id
+  InternedCache<std::pair<uint64_t, uint64_t>> interned_callstack_id_cache_;
+  InternedCache<std::pair<uint64_t, uint64_t>> interned_string_id_cache_;
+  InternedCache<std::pair<uint64_t, uint64_t>> interned_tracepoint_id_cache_;
+};
+
+void ProducerEventProcessorImpl::ProcessAddressInfo(uint64_t producer_id,
+                                                    AddressInfo* address_info) {
+  // TODO: Split address info into 2 separate messages, one with strings another one with keys.
+  if (address_info->function_name_or_key_case() == AddressInfo::kFunctionName) {
+    auto [function_name_key, assigned] = string_cache_.GetOrAssignId(address_info->function_name());
+    if (assigned) {
+      SendStringAndKeyEvent(function_name_key, address_info->function_name());
+    }
+    address_info->set_function_name_key(function_name_key);
+  } else {
+    CHECK(address_info->function_name_or_key_case() == AddressInfo::kFunctionNameKey);
+    auto [function_name_key, assigned] =
+        interned_string_id_cache_.GetOrAssignId({producer_id, address_info->function_name_key()});
+    CHECK(!assigned);
+    address_info->set_function_name_key(function_name_key);
+  }
+
+  if (address_info->map_name_or_key_case() == AddressInfo::kMapName) {
+    auto [module_name_key, assigned] = string_cache_.GetOrAssignId(address_info->map_name());
+    if (assigned) {
+      SendStringAndKeyEvent(module_name_key, address_info->map_name());
+    }
+    address_info->set_map_name_key(module_name_key);
+  } else {
+    CHECK(address_info->map_name_or_key_case() == AddressInfo::kMapNameKey);
+    auto [module_name_key, assigned] =
+        interned_string_id_cache_.GetOrAssignId({producer_id, address_info->map_name_key()});
+    CHECK(!assigned);
+    address_info->set_map_name_key(module_name_key);
+  }
+
+  ClientCaptureEvent event;
+  *event.mutable_address_info() = std::move(*address_info);
+  capture_event_buffer_->AddEvent(std::move(event));
+}
+
+void ProducerEventProcessorImpl::ProcessFunctionCall(FunctionCall* function_call) {
+  ClientCaptureEvent event;
+  *event.mutable_function_call() = std::move(*function_call);
+  capture_event_buffer_->AddEvent(std::move(event));
+}
+
+void ProducerEventProcessorImpl::ProcessGpuJob(uint64_t producer_id, GpuJob* gpu_job) {
+  // TODO: Split GpuJob into 2 separate messages, one with string and another one with the key
```

Comment:
Again, either add a bug id or address this now (or also fine for me, to address this right after this PR).